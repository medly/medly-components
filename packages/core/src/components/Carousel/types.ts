import { BreakpointsTheme } from '@medly-components/theme';

export type StyledNextPrevBtnProps = {
    disabled?: boolean;
    top?: string;
};

export type CarouselStyledProps = {
    itemsCount?: {
        [k in keyof BreakpointsTheme]: number;
    };
    arrowPositionFromTop?: string;
};

export type CarouselProps = CarouselStyledProps & {
    withVariableWidth?: boolean;
};
