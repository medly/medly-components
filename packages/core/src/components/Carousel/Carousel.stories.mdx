import { Preview, Story, Meta, Props } from '@storybook/addon-docs/blocks';
import { withKnobs } from '@storybook/addon-knobs';
import { defaultTheme } from '@medly-components/theme';
import { Carousel } from './Carousel.tsx';
import { useState } from 'react';

<Meta title="Core" component={Carousel} parameters={{ jest: ['Carousel.test.tsx'] }} />

# Carousel

<Preview withToolbar>
    <Story name="Accordion">
        <Carousel></Carousel>
    </Story>
</Preview>

## Props

This component offers the following props mentioned in the table below:

<Props of={Carousel} />
