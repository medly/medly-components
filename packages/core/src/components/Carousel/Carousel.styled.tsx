import { ArrowLeftIcon, ArrowRightIcon } from '@medly-components/icons';
import { BreakpointsTheme } from '@medly-components/theme';
import { breakpoints, media } from '@medly-components/utils';
import styled, { css } from 'styled-components';
import { CarouselStyledProps, StyledNextPrevBtnProps } from './types';

const commonStyle = css`
    position: absolute;
    cursor: pointer;
    transform: translateY(-50%);
    display: none;

    ${({ theme }) => media(breakpoints(theme.breakpoints).up('L'))`
        display: block;
    `}
`;

export const CarouselWrapper = styled.div`
    position: relative;
`;

export const CarouselScroll = styled.ul<CarouselStyledProps>`
    list-style: none;
    padding: 0;
    margin: 0;
    overflow-x: scroll;
    overflow-y: hidden;
    display: flex;
    align-items: center;
    scrollbar-width: none;
    ::-webkit-scrollbar {
        display: none;
    }

    ${({ theme, itemsCount }) =>
        itemsCount &&
        css`
            ${Object.entries(itemsCount)?.reduce(
                (acc, [key, value]) =>
                    `${acc}
                            ${breakpoints(theme.breakpoints).up(key as keyof BreakpointsTheme)}{
                                & > * {
                                    min-width: calc(100% / ${value});
                                }
                            }
                        `,
                ''
            )}
        `}
`;

export const LeftArrowIcon = styled(ArrowLeftIcon)<StyledNextPrevBtnProps>`
    ${commonStyle}
    left: -5.6rem;
    top: ${({ top }) => top || '50%'};
    visibility: ${({ disabled }) => disabled && 'hidden'};
    ${({ theme }) => media(breakpoints(theme.breakpoints).down('S'))`
        left: 0.6rem;
    `}
`;

export const RightArrowIcon = styled(ArrowRightIcon)<StyledNextPrevBtnProps>`
    ${commonStyle}
    right: -5.6rem;
    top: ${({ top }) => top || '50%'};
    visibility: ${({ disabled }) => disabled && 'hidden'};
    ${({ theme }) => media(breakpoints(theme.breakpoints).down('S'))`
        right: 0.6rem;
    `}
`;
