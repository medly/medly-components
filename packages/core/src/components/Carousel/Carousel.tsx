import { FC, useEffect, useRef, useState } from 'react';
import { CarouselScroll, CarouselWrapper, LeftArrowIcon, RightArrowIcon } from './Carousel.styled';
import { CarouselProps } from './types';

export const Carousel: FC<CarouselProps> = ({ children, itemsCount, withVariableWidth, arrowPositionFromTop, ...restProps }) => {
    const carouselRef = useRef<HTMLUListElement>(null),
        [isScrolledToLeft, setIsScrolledToLeft] = useState(true),
        [isScrolledToRight, setIsScrolledToRight] = useState(false);

    const getScrollWidth = () =>
            carouselRef.current
                ? withVariableWidth
                    ? carouselRef.current?.clientWidth * 0.4
                    : carouselRef.current?.children[0].clientWidth
                : 0,
        scrollNext = () => carouselRef.current?.scrollBy({ left: getScrollWidth(), behavior: 'smooth' }),
        scrollPrev = () => carouselRef.current?.scrollBy({ left: -getScrollWidth(), behavior: 'smooth' });

    const handleScroll = () => {
        if (carouselRef.current) {
            setIsScrolledToLeft(carouselRef.current.scrollLeft === 0);
            setIsScrolledToRight(carouselRef.current.scrollWidth - (carouselRef.current.scrollLeft + carouselRef.current.clientWidth) <= 1);
        }
    };

    useEffect(() => {
        carouselRef.current && setIsScrolledToRight(carouselRef.current?.clientWidth === carouselRef.current?.scrollWidth);
    }, [carouselRef.current?.scrollWidth]);

    return (
        <CarouselWrapper {...restProps}>
            <LeftArrowIcon
                disabled={isScrolledToLeft}
                data-testid="carousel-left-arrow-icon"
                onClick={scrollPrev}
                top={arrowPositionFromTop}
            />
            <CarouselScroll ref={carouselRef} data-testid="carousel-scroll" onScroll={handleScroll} itemsCount={itemsCount}>
                {children}
            </CarouselScroll>
            <RightArrowIcon
                data-testid="carousel-right-arrow-icon"
                disabled={isScrolledToRight}
                onClick={scrollNext}
                top={arrowPositionFromTop}
            />
        </CarouselWrapper>
    );
};
