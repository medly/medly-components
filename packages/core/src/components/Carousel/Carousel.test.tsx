import { fireEvent, render, screen } from '@test-utils';
import { Carousel } from './Carousel';

describe('Carousel', () => {
    it('should render properly with the children of variable width', () => {
        const { container } = render(
            <Carousel
                itemsCount={{
                    S: 2.75,
                    M: 3.75,
                    L: 6,
                    XL: 6
                }}
                withVariableWidth
            >
                <span>Hello Dummy Text</span>
                <span>Hello Dummy</span>
            </Carousel>
        );
        expect(container).toMatchSnapshot();
    });

    it('should render properly with the children of same width', () => {
        const { container } = render(
            <Carousel
                itemsCount={{
                    S: 2.75,
                    M: 3.75,
                    L: 6,
                    XL: 6
                }}
            >
                <span>Hello Dummy</span>
                <span>Hello Dummy</span>
            </Carousel>
        );
        expect(container).toMatchSnapshot();
    });

    it('should make the left arrow disabled on clicking on left arrow', () => {
        render(
            <Carousel
                itemsCount={{
                    S: 2,
                    M: 2,
                    L: 2,
                    XL: 2
                }}
            >
                <span>Hello Dummy</span>
                <span>Hello Dummy</span>
                <span>Hello Dummy</span>
                <span>Hello Dummy</span>
                <span>Hello Dummy</span>
                <span>Hello Dummy</span>
                <span>Hello Dummy</span>
            </Carousel>
        );
        const leftArrow = screen.getByTestId('carousel-left-arrow-icon');
        const element = screen.getByTestId('carousel-scroll');
        expect(leftArrow).toBeDisabled();
        fireEvent.scroll(element, { target: { scrollLeft: 100 } });
        expect(leftArrow).toBeEnabled();
    });

    it('should make the right arrow disabled on clicking on right arrow', () => {
        render(
            <Carousel
                itemsCount={{
                    S: 2,
                    M: 2,
                    L: 2,
                    XL: 2
                }}
            >
                <span>Hello Dummy</span>
                <span>Hello Dummy</span>
                <span>Hello Dummy</span>
                <span>Hello Dummy</span>
                <span>Hello Dummy</span>
                <span>Hello Dummy</span>
                <span>Hello Dummy</span>
            </Carousel>
        );
        const rightArrow = screen.getByTestId('carousel-right-arrow-icon');
        const element = screen.getByTestId('carousel-scroll');
        expect(rightArrow).toBeDisabled();
        fireEvent.scroll(element, { target: { scrollLeft: -10 } });
        expect(rightArrow).toBeEnabled();
    });
});
