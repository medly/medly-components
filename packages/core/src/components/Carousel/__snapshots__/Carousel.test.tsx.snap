// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel should render properly with the children of same width 1`] = `
.c0 {
  position: relative;
}

.c2 {
  list-style: none;
  padding: 0;
  margin: 0;
  overflow-x: scroll;
  overflow-y: hidden;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-scrollbar-width: none;
  -moz-scrollbar-width: none;
  -ms-scrollbar-width: none;
  scrollbar-width: none;
}

.c2::-webkit-scrollbar {
  display: none;
}

.c1 {
  position: absolute;
  cursor: pointer;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
  display: none;
  left: -5.6rem;
  top: 50%;
  visibility: hidden;
}

.c3 {
  position: absolute;
  cursor: pointer;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
  display: none;
  right: -5.6rem;
  top: 50%;
  visibility: hidden;
}

@media (min-width:0px) {
  .c2 > * {
    min-width: calc(100% / 2.75);
  }
}

@media (min-width:768px) {
  .c2 > * {
    min-width: calc(100% / 3.75);
  }
}

@media (min-width:1024px) {
  .c2 > * {
    min-width: calc(100% / 6);
  }
}

@media (min-width:1280px) {
  .c2 > * {
    min-width: calc(100% / 6);
  }
}

@media (min-width:1024px) {
  .c1 {
    display: block;
  }
}

@media (max-width:767px) {
  .c1 {
    left: 0.6rem;
  }
}

@media (min-width:1024px) {

}

@media (max-width:767px) {

}

@media (min-width:1024px) {
  .c3 {
    display: block;
  }
}

@media (max-width:767px) {
  .c3 {
    right: 0.6rem;
  }
}

<div>
  <div
    class="c0"
  >
    <svg
      class="c1"
      data-testid="carousel-left-arrow-icon"
      disabled=""
      fill="#303C48"
      height="32"
      viewBox="0 0 256 256"
      width="32"
      xmlns="http://www.w3.org/2000/svg"
    >
      <rect
        fill="none"
        height="256"
        width="256"
      />
      <polyline
        fill="none"
        points="160 208 80 128 160 48"
        stroke="#303C48"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="16"
      />
    </svg>
    <ul
      class="c2"
      data-testid="carousel-scroll"
    >
      <span>
        Hello Dummy
      </span>
      <span>
        Hello Dummy
      </span>
    </ul>
    <svg
      class="c3"
      data-testid="carousel-right-arrow-icon"
      disabled=""
      fill="#303C48"
      height="32"
      viewBox="0 0 256 256"
      width="32"
      xmlns="http://www.w3.org/2000/svg"
    >
      <rect
        fill="none"
        height="256"
        width="256"
      />
      <polyline
        fill="none"
        points="96 48 176 128 96 208"
        stroke="#303C48"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="16"
      />
    </svg>
  </div>
</div>
`;

exports[`Carousel should render properly with the children of variable width 1`] = `
.c0 {
  position: relative;
}

.c2 {
  list-style: none;
  padding: 0;
  margin: 0;
  overflow-x: scroll;
  overflow-y: hidden;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-scrollbar-width: none;
  -moz-scrollbar-width: none;
  -ms-scrollbar-width: none;
  scrollbar-width: none;
}

.c2::-webkit-scrollbar {
  display: none;
}

.c1 {
  position: absolute;
  cursor: pointer;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
  display: none;
  left: -5.6rem;
  top: 50%;
  visibility: hidden;
}

.c3 {
  position: absolute;
  cursor: pointer;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
  display: none;
  right: -5.6rem;
  top: 50%;
  visibility: hidden;
}

@media (min-width:0px) {
  .c2 > * {
    min-width: calc(100% / 2.75);
  }
}

@media (min-width:768px) {
  .c2 > * {
    min-width: calc(100% / 3.75);
  }
}

@media (min-width:1024px) {
  .c2 > * {
    min-width: calc(100% / 6);
  }
}

@media (min-width:1280px) {
  .c2 > * {
    min-width: calc(100% / 6);
  }
}

@media (min-width:1024px) {
  .c1 {
    display: block;
  }
}

@media (max-width:767px) {
  .c1 {
    left: 0.6rem;
  }
}

@media (min-width:1024px) {

}

@media (max-width:767px) {

}

@media (min-width:1024px) {
  .c3 {
    display: block;
  }
}

@media (max-width:767px) {
  .c3 {
    right: 0.6rem;
  }
}

<div>
  <div
    class="c0"
  >
    <svg
      class="c1"
      data-testid="carousel-left-arrow-icon"
      disabled=""
      fill="#303C48"
      height="32"
      viewBox="0 0 256 256"
      width="32"
      xmlns="http://www.w3.org/2000/svg"
    >
      <rect
        fill="none"
        height="256"
        width="256"
      />
      <polyline
        fill="none"
        points="160 208 80 128 160 48"
        stroke="#303C48"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="16"
      />
    </svg>
    <ul
      class="c2"
      data-testid="carousel-scroll"
    >
      <span>
        Hello Dummy Text
      </span>
      <span>
        Hello Dummy
      </span>
    </ul>
    <svg
      class="c3"
      data-testid="carousel-right-arrow-icon"
      disabled=""
      fill="#303C48"
      height="32"
      viewBox="0 0 256 256"
      width="32"
      xmlns="http://www.w3.org/2000/svg"
    >
      <rect
        fill="none"
        height="256"
        width="256"
      />
      <polyline
        fill="none"
        points="96 48 176 128 96 208"
        stroke="#303C48"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="16"
      />
    </svg>
  </div>
</div>
`;
